### PR TITLE
Add option to reserve/release testbeds in downstream lab

### PR DESCRIPTION
**Pull request by vasukulkarni**
Fixes: https://github.com/red-hat-storage/ocs-ci/issues/1999

Signed-off-by: Vasu Kulkarni <vasu@redhat.com>

### CODE REVIEW

**Comment by vasukulkarni** [comment]
Verified the reservation which seems to be working fine, I just added 5 slots for Common pool for now

```
run-ci --reserve --testbed-name Common -m deployment --deploy tests/ --cluster-name vasukulkarni --cluster-path ~/my-cluster-path
2020-08-17 10:10:16,506 - MainThread - ocs_ci.utility.utils - INFO - Retrieving the authentication config dictionary
2020-08-17 10:10:16,507 - MainThread - /home/vasu/ocs-ci/ocs_ci/utility/aws.py - INFO - Fetching authentication credentials from ocs-ci-data
2020-08-17 10:10:16,518 - MainThread - botocore.credentials - INFO - Found credentials in shared credentials file: ~/.aws/credentials
2020-08-17 10:10:17,969 - MainThread - ocs_ci.ocs.openshift_ops - INFO - Testing access to cluster with /home/vasu/my-cluster-path/auth/kubeconfig
2020-08-17 10:10:17,969 - MainThread - ocs_ci.ocs.openshift_ops - WARNING - The kubeconfig file /home/vasu/my-cluster-path/auth/kubeconfig doesn't exist!
2020-08-17 10:10:17,970 - MainThread - ocs_ci.utility.utils - INFO - testrun_name: OCS4-5-Downstream-OCP4-5-AWS-IPI-1AZ-RHCOS-3M-3W-deployment
2020-08-17 10:10:18,223 - MainThread - ocs_ci.framework.testbed - INFO - No Existing reservation found
2020-08-17 10:10:18,463 - MainThread - ocs_ci.framework.testbed - INFO - 201
2020-08-17 10:10:18,464 - MainThread - ocs_ci.framework.testbed - INFO - Post is successful
2020-08-17 10:10:18,676 - MainThread - ocs_ci.framework.testbed - INFO - Successfuly sent the request
2020-08-17 10:10:18,676 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - (False, 'Common', None)
2020-08-17 10:10:18,676 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - Waiting for testbed to be reserved
2020-08-17 10:10:18,676 - MainThread - ocs_ci.utility.utils - INFO - Going to sleep for 60 seconds before next iteration
2020-08-17 10:11:18,978 - MainThread - ocs_ci.framework.testbed - INFO - Successfuly sent the request
2020-08-17 10:11:18,978 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - (True, 'Cyanea_Common_DC1_C1', {'s3': 's3://ocs-ci-data/vSphere-DC-CP_VC1.yaml', 'gdrive': 'None'})
2020-08-17 10:11:18,978 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - Successfully Reserved Cyanea_Common_DC1_C1
2020-08-17 10:11:18,979 - MainThread - /home/vasu/ocs-ci/ocs_ci/utility/aws.py - INFO - Fetching authentication credentials from ocs-ci-data
/home/vasu/ocs-ci/ocs_ci/framework/pytest_customization/ocscilib.py:370: ResourceWarning: unclosed <ssl.SSLSocket fd=13, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.1.197', 32874), raddr=('52.219.100.40', 443)>
  update_config_from_s3(bucket_name, testbed_name)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
2020-08-17 10:11:28,225 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - Dump of the consolidated config file is located here: /tmp/run-1597684216-config.yaml
2020-08-17 10:11:28,225 - MainThread - ocs_ci.framework.pytest_customization.ocscilib - INFO - Skipping versions collecting because: Deploy or destroy of cluster is performed.
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.8.5, pytest-5.3.5, py-1.9.0, pluggy-0.13.1
rootdir: /home/vasu/ocs-ci, inifile: pytest.ini
plugins: logger-0.5.1, ordering-0.6, html-2.1.1, reportportal-1.10.0, marker-bugzilla-0.9.4, metadata-1.10.0
collecting ... 2020-08-17 10:11:28,608 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  --selector=node-role.kubernetes.io/worker -o yaml

------------------------------------------------------------------------------------------------------------ live log collection -------------------------------------------------------------------------------------------------------------
10:11:28 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc get Node  --selector=node-role.kubernetes.io/worker -o yaml
10:11:28 - MainThread - ocs_ci.utility.utils - WARNING - Command stderr: error: Missing or incomplete configuration info.  Please point to an existing, complete config file:

```

File: tests/conftest.py [hunk 1]
```diff
@@ -1045,6 +1046,7 @@ def cluster(request, log_cli_level):
     if teardown:
         def cluster_teardown_finalizer():
             deployer.destroy_cluster(log_cli_level)
+            release_testbed(config.RUN['testbed_name'])
```
Comment (vasukulkarni) [inline tests/conftest.py:1049]: have to check if we have already reserved!

**Review by clacroix12 — COMMENTED** [review]
Just a few small (non functional) things, nothing critical that should hold the PR up though.

File: ocs_ci/framework/testbed.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    pass
+
+
+class TestbedReleaseFailed(Exception):
+    pass
+
+
+def reserve_testbed(launch_name, user, testbed_name, timeout):
+    """
+    Function to reserve the testbed
```
Comment (ebenahar) [inline ocs_ci/framework/testbed.py:27]: what does `testbed` represent?

File: ocs_ci/framework/testbed.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
+import logging
```
Comment (ebenahar) [inline ocs_ci/framework/testbed.py:1]: Should we put the file under ocs_ci/framework/? the content of this file doesn't match with the other files under this dir

**Comment by petr-balogh** [comment]
Can we close @vasukulkarni ?

**Pull request closed without merging**